File: bookshelf-service/internal/domain/models/book.go
// Package models contains the domain models of the application
package models

// Book represents a book in the application and database
type Book struct {
	ID     string `json:"id" gorm:"primarykey"`
	UserID string `json:"-"`
	Title  string `json:"title"`
	Author string `json:"author"`
	Genre  string `json:"genre"`
	Read   bool   `json:"read"`
	Rating int    `json:"rating,omitempty"`
}

File: bookshelf_app/lib/data/models/book.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String? genre;
  final int? rating;
  final bool read;

  Book({
    required this.id,
    required this.title,
    required this.author,
    this.genre,
    this.rating,
    this.read = false,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'author': author,
        'genre': genre,
        'rating': rating,
        'read': read,
      };

  factory Book.fromJson(Map<String, dynamic> json) => Book(
        id: json['id'],
        title: json['title'],
        author: json['author'],
        genre: json['genre'],
        rating: json['rating'],
        read: json['read'] ?? false,
      );

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Book && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;
}
